### PR TITLE
Improve docs indexing, accessibility, and initial load performance

### PR DESCRIPTION
**Pull request by Alek99**
## Problem and result

Mounted production builds were overwriting prerendered docs with asset directories and the root SPA shell. This left the docs home without indexable content and six articles returning 404. Other pages repeated metadata, exposed legacy URLs as 200/JavaScript redirects, and loaded substantial code and chart work before readers needed it.

This PR repairs the production output, improves shared docs structure and accessibility, and reduces initial work while preserving interactive examples. The complete final production crawl verifies **418/418 canonical pages**, and all **94 legacy redirects** resolve correctly.

## What changed and why

- **Preserve mounted prerenders:** recursively merge assets into the route tree and compress only after the final tree is assembled, so routes named assets/components and the docs home survive the build.
- **Make canonical pages distinct and discoverable:** contextual titles/descriptions with normalized standalone acronyms, canonical URLs and structured breadcrumbs using the configured deployment origin/mount and real routes (localhost by default; `REFLEX_DEPLOY_URL` supplied by deployment jobs), the standard sitemap XML namespace, and frontend-prefix-aware sitemap generation. Fix duplicate H1 content, broken fragments, and missing image alt text.
- **Return real redirects when the backend serves the frontend:** GET/HEAD middleware issues permanent 301 responses and retains queries. Static-only hosting gets readable destination links, canonical/noindex metadata, and refresh fallback; CDN edge redirects still need deployment configuration.
- **Render readable code immediately:** escaped Shiki fallback code is present in SSR. Download/highlight near the viewport during idle time; failed highlighting leaves code readable. Cached HTML matches every highlighting input, so theme/language/transformer changes immediately restore fallback code until replacement highlighting succeeds. Preserve cleanup on unmount.
- **Reduce initial library work:** standard React memo initializers can be discarded when unused, while custom wrappers retain side effects. Add the opt-in `frontend_lazy_bundled_libraries` configuration and enable it for docs; optional dynamic namespaces load once on demand and retry after failures. Core React/context identity is preserved, and the framework default stays eager for compatibility. Preload global CSS explicitly.
- **Defer expensive examples:** the documented `demo exec defer` flag mounts Plotly previews as readers approach them, reserves space, and preserves server-rendered source. Charts now fit their mobile container. Remove docs-wide marketing pixels/session recorders.
- **Improve shared controls/layout:** descriptions render once in responsive API tables; copy buttons, card links, clear controls, mobile navigation, and feedback triggers have valid semantics/names. Eliminate nested feedback buttons, keep card focus outlines inside rounded clipping boundaries, and improve code/counter/status contrast.
- **Isolate project configuration:** only the requested project can supply rxconfig, avoiding an unrelated editable docs app changing framework tests or another project. Adjust the invalid-event-value fixture to account for the enterprise plugin's legitimate function serializer. Scope version-check mocks to their consumer so lazy dependency imports still read real package versions. The docs link validator uses the standard sitemap namespace and the configured build directory.

## Before and after

| Check | Before | After |
|---|---|---|
| Docs home | Empty 14.8 KiB SPA shell | Full prerendered content |
| Broken articles | 6 return 404 | All 6 return 200 |
| Canonical crawl | Missing home metadata; duplicate titles/descriptions | 418 unique titles/descriptions, one H1 each, matching canonicals |
| Legacy routes | 94 return 200 with JS redirects | 94 HTTP 301s; all destinations and static fallbacks verified |
| Sitemap | Invalid namespace / wrong mounted root | XSD-valid, 418 unique production URLs including docs home |
| Broken in-page fragments / missing image alt attributes | 2 / 4 | 0 / 0 |
| Structured breadcrumbs | None | 417 articles |
| Plotly mobile initial total blocking time | 5,965 ms | 0 ms |

Lighthouse performance scores (out of 100), **before → after**:

| Page | Mobile | Desktop |
|---|---:|---:|
| Docs home | 63 → 68 | 89 → 97 |
| Introduction | 53 → 57 | 81 → 85 |
| Installation | 47 → 60 | 80 → 86 |
| Chat tutorial | 55 → 59 | 87 → 89 |
| Button | 55 → 57 | 75 → 81 |
| Bar chart | 42 → 53 | 78 → 80 |
| AG Grid | 44 → 49 | 85 → 79 |
| App API | 54 → 60 | 81 → 89 |
| Changelog | 57 → 61 | 88 → 92 |
| HTML layout | 49 → 57 | 77 → 83 |
| Plotly | 23 → 53 | 67 → 81 |

Lighthouse SEO is **100 in all 26 recorded runs**. Best practices improved from 54 to 100 except AG Grid (96, due to enterprise trial-license console messages). SEO scores alone missed the original broken/empty pages, so the full raw-HTML crawl is the primary correctness check.

Initial local JavaScript in the HTML/module-preload graph (gzip):

| Route | Before → after |
|---|---:|
| `/docs/getting-started/introduction/` | 889 → 798 KiB |
| `/docs/library/forms/button/` | 1070 → 960 KiB |
| `/docs/api-reference/app/` | 586 → 488 KiB |
| `/docs/library/graphing/other-charts/plotly/` | 1076 → 966 KiB |

**Measurement provenance:** the Lighthouse table was recorded on `654132a`. The subsequent review fixes change highlighting invalidation, focus outlines, title casing, breadcrumb edge cases, and test fixtures; all 26 performance runs were not repeated after those fixes. The complete 418-page crawl and production browser checks were repeated at `fe56d43a`. The follow-up in `f2a0f434` was checked with the complete docs suite, all four frontend regressions, real app imports under staging URL settings, and separate dev/static frontend redirect checks.

**Method:** production builds against base `bae7294119a8ffa2b3177a0a0f0063d76b616d86`, same macOS/Apple M5 Pro machine, Node 26, Chrome 152, Lighthouse 13.4.1 with default simulated throttling and fresh headless Chrome launches. Eleven routes on mobile/desktop; home/introduction mobile scores are medians of three runs, others single runs. All 26 results are included, including regressions. These local measurements are not field Core Web Vitals. The home originally served an empty shell, so its corrected payload is not an equivalent isolated speed comparison.

## Validation

- **8,453 framework unit tests passed**, 18 skipped; **76.50% coverage** collected from process startup with `uv run --no-sync coverage run -m pytest tests/units -q`, then `coverage report`.
- **449 docs tests passed**, 1 skipped, from `docs/app`.
- The `reflex-docs` Python 3.11/3.12 CI jobs now run the frontend suite after the production build, so these checks run on each PR.
- **4 frontend regressions passed:** escaped React SSR, highlight input invalidation/failure/async replacement, Rolldown removal of unused memos/custom-wrapper side effects, and generated lazy-registry execution/concurrency/retry/legacy compatibility.
- All changed-file pre-commit hooks passed: Ruff formatting/lint, codespell, generated stubs, Pyright, ty, Biome. Full-repository Ruff checks and `pyright reflex tests` pass; no stub hash changes.
- Complete production build, 418-page HTML crawl, 94 HTTP/static redirects, sitemap XSD validation, and 26 Lighthouse runs.
- Browser checks: counter updates, feedback popover with a single submit button, mobile navigation, readable light/dark code, wrapping API rows, and one deferred Plotly chart measuring 308px inside a 308px parent with no page overflow at 390px.

Initial validation coverage detail (654132a): pytest-cov's standard invocation passes all unit tests but reports 71.14% because optional pytest plugins import framework modules before its hook starts. Starting coverage before pytest measures 76.50%; the separate units-plus-docs pytest-cov run measures 72.90%. Earlier CLI failures were resolved by fetching Git history/tags and reinstalling editable packages to replace invalid shallow-clone version metadata. No assertions or coverage thresholds were disabled.

## Review notes and limits

1. AG Grid desktop is 85 → 79 in the final single-run comparison; mobile is 44 → 49. Shared providers, CSS, and example bundles remain substantial. This PR does not claim every page is maximally fast.
2. Some demo-specific accessibility/table checks and vendor trial warnings remain. No audit checks or vendor errors are suppressed.
3. Verify redirects/cache behavior in the real hosting stack. The separate XY docs application is not part of this checkout and remains the one missing local cross-application target.
4. Broad integration and minimum-dependency matrices were not run locally. Unit/docs coverage and production browser verification are complete; the repository CI matrix remains the release gate.
5. Related open work: #6723 and #6732 overlap accessible controls and code loading (the latter uses build-time Shiki); #6563 addresses namespace tree-shaking with a different approach; #7075 changes same-root config dependency eviction. This PR is based on current main and does not include those branches. Coordinate overlapping changes during review.

### Submission checklist

- [x] Reviewed CONTRIBUTING.md and checked related open PRs.
- [x] Bug fixes and opt-in performance behavior documented.
- [x] Regression assertions added; tests and pre-commit checks run.
- [x] News fragments included for each affected package.
- [x] Adversarial review completed; limitations and overlap recorded above.

<a href="https://cubic.dev/pr/reflex-dev/reflex/pull/7078?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

The subsequent `a7e412af` follow-up removes the hardcoded production origin from docs config, preserves APIs/URLs casing in page titles and sidebar labels, and passes 449 docs tests (1 skipped). The `show_built_with_reflex=False` setting remains enabled in the docs config to hide the badge.

### CODE REVIEW

File: packages/reflex-base/src/reflex_base/.templates/web/components/shiki/code.js [hunk 1]
```diff
@@ -1,40 +1,89 @@
...
-  });
+    return () => {
+      active = false;
+      observer?.disconnect();
+      if ("cancelIdleCallback" in window) window.cancelIdleCallback(idle);
+      else window.clearTimeout(idle);
+    };
+  }, [code, language, theme, themes, transformers, decorations]);
+
+  if (highlighted?.code === code) {
```
[the remark this answers is not in the document]
Comment (Alek99) [reply packages/reflex-base/src/reflex_base/.templates/web/components/shiki/code.js]: Fixed in fe56d43a. Cached HTML now matches code, language, theme/themes, transformers, and decorations before it can render. A changed input immediately renders readable fallback code, including when replacement highlighting fails. The new executable component regression covers all six inputs, failed replacements, and an older async result finishing after its replacement.

[the remark this answers is not in the document]
Comment (Alek99) [reply packages/reflex-base/src/reflex_base/.templates/web/components/shiki/code.js]: Fixed in fe56d43a. Cached HTML now matches code, language, theme/themes, transformers, and decorations before it can render. A changed input immediately renders readable fallback code, including when replacement highlighting fails. The new executable component regression covers all six inputs, failed replacements, and an older async result finishing after its replacement.

File: docs/app/reflex_docs/templates/docpage/docpage.py [hunk 2]
```diff
@@ -182,6 +182,53 @@ def docpage_footer(path: rx.Var[str], edit_href: rx.Var[str]) -> rx.Component:
...
+            "position": len(items) + 1,
+            "name": _BREADCRUMB_LABEL_OVERRIDES.get(label, label),
+            "item": base + href,
+        })
+        seen.add(base + href)
+    items.append({
+        "@type": "ListItem",
+        "position": len(items) + 1,
+        "name": title,
+        "item": canonical,
```
[the remark this answers is not in the document]
Comment (Alek99) [reply docs/app/reflex_docs/templates/docpage/docpage.py]: Fixed in fe56d43a. The breadcrumb helper now emits the documentation root once. Regression cases cover both `/` and the empty normalized root; the existing nested-route assertions still pass.

File: docs/app/reflex_docs/pages/docs/metadata.py [hunk 3]
```diff
@@ -19,3 +19,43 @@ def truncate_meta_description(description: str, max_len: int = 155) -> str:
...
+    ]
+    if title.lower() == "index" and parents:
+        title = " ".join(
+            acronyms.get(word, word.capitalize())
+            for word in path.strip("/").split("/")[-1].split("-")
+        )
+    context = [
+        parent for parent in reversed(parents) if parent.lower() != title.lower()
+    ]
+    subject = " · ".join([title, *context])
```
[the remark this answers is not in the document]
Comment (Alek99) [reply docs/app/reflex_docs/pages/docs/metadata.py:57]: Fixed in fe56d43a. Standalone acronym words in titles now use the acronym map. Tests cover `Cli` → `CLI`, `Api Reference` → `API Reference`, and preservation of the code identifier `rx.html`.

File: tests/units/components/test_memo.py [hunk 4]
```diff
@@ -2321,3 +2321,19 @@ class _BetaProbe(Component):
...
+    )
+
+    @rx.memo(wrapper=wrapper)
+    def tracked_named_memo() -> rx.Component:
+        return rx.text("Tracked")
+
+    definition = MEMOS["TrackedNamedMemo", __name__]
+    files, _ = compiler.compile_memo_components((definition,))
+    code = "\n".join(content for _, content in files)
+    assert "/*#__PURE__*/" not in code
```
[the remark this answers is not in the document]
Comment (Alek99) [reply tests/units/components/test_memo.py:2340]: Added the positive control in fe56d43a: `test_component_memo_default_wrapper` now asserts the default React wrapper emits the PURE initializer. The custom wrapper named `memo` still asserts the annotation is absent, so removing the gate entirely can no longer satisfy both cases.

File: docs/app/reflex_docs/pages/docs_landing/views/ai_builder.py [hunk 5]
```diff
@@ -88,7 +88,11 @@ def card(
             class_name="flex flex-col gap-2 p-8",
         ),
         content,
-        rx.el.a(href=href, class_name="absolute inset-0"),
+        rx.el.a(
+            href=href,
+            aria_label=title,
+            class_name="absolute inset-0 rounded-xl focus-visible:outline-2 focus-visible:outline-offset-4 focus-visible:outline-primary-9",
```
[the remark this answers is not in the document]
Comment (Alek99) [reply docs/app/reflex_docs/pages/docs_landing/views/ai_builder.py]: Fixed in fe56d43a. The outline now uses a negative offset, placing the 2px focus ring inside the rounded card's clipping boundary. The regression checks the inward offset, and the production keyboard-focus check confirms the ring remains visible.

File: docs/app/tests/frontend_quality.test.mjs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+const resolve = (name) => pathToFileURL(require.resolve(name)).href;
+
+// Use the exact React and production bundler versions installed by Reflex.
+test("Shiki server rendering includes escaped, readable code", async () => {
+  const { createElement } = await import(resolve("react"));
+  const { renderToStaticMarkup } = await import(resolve("react-dom/server"));
+  const filename = path.join(root, "packages/reflex-base/src/reflex_base/.templates/web/components/shiki/code.js");
+  const source = (await readFile(filename, "utf8"))
+    .replaceAll('from "react"', `from "${resolve("react")}"`)
+    .replaceAll('from "shiki"', `from "${resolve("shiki")}"`);
```
[the remark this answers is not in the document]
Comment (Alek99) [reply docs/app/tests/frontend_quality.test.mjs]: Removed the unused replacement in fe56d43a. The SSR test now explicitly states that it checks React's escaped fallback and does not run Shiki effects. The new lifecycle regression replaces the actual dynamic import form with a controlled highlighter to exercise success, failure, and stale work.

File: docs/app/tests/frontend_quality.test.mjs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    bundle_library('quality-lazy-fixture')
+    print(compile_app_root(rx.el.div('Quality'))[1])
+`], {cwd: root, encoding: "utf8"});
+  const setup = new Function("window", "React", "emotion_react", "utils_context", "utils_state", "loadFixture",
+    rootCode.replace(/^import .*$/gm, "")
+      .replace(/^export default function/gm, "function")
+      .replace(/^export /gm, "")
+      .replace('import("quality-lazy-fixture")', 'loadFixture()'));
+  const stateSource = await readFile(path.join(root, "packages/reflex-base/src/reflex_base/.templates/web/utils/state.js"), "utf8");
+  const evaluateSource = stateSource.split("export const evalReactComponent = ")[1].split("\n};")[0] + "\n}";
```
[the remark this answers is not in the document]
Comment (Alek99) [reply docs/app/tests/frontend_quality.test.mjs]: Replaced the delimiter splitting in fe56d43a. The test uses the installed Rolldown module parser to locate the complete named export, then loads it as an ES module. Nested statements and braces no longer determine the extraction boundary, and the concurrency/retry/legacy execution checks still pass.

File: docs/app/tests/frontend_quality.test.mjs [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+      .replace(/^export /gm, "")
+      .replace('import("quality-lazy-fixture")', 'loadFixture()'));
+  const stateSource = await readFile(path.join(root, "packages/reflex-base/src/reflex_base/.templates/web/utils/state.js"), "utf8");
+  // Parse the complete module so nested statements cannot truncate the export.
+  const { parseAst } = await import(resolve("rolldown/parseAst"));
+  const stateModule = parseAst(stateSource);
+  const evaluateExport = stateModule.body.find(node => node.type === "ExportNamedDeclaration" &&
+    node.declaration?.declarations?.some(declaration => declaration.id.name === "evalReactComponent"));
+  assert(evaluateExport, "state.js exports evalReactComponent");
+  const { evalReactComponent: evaluate } = await import(moduleUrl(stateSource.slice(evaluateExport.start, evaluateExport.end)));
```
[the remark this answers is not in the document]
Comment (Alek99) [reply docs/app/tests/frontend_quality.test.mjs:162]: Verified this against the installed Rolldown 1.2.6 parser; this finding does not apply. Its JavaScript AST exposes UTF-16 string offsets, not the underlying Rust UTF-8 byte spans. For a prefix `// café 🔥 日本語\n`, `parseAst` returns `start: 15`, matching the JavaScript string length; the UTF-8 byte length is 24. The existing string slice returns the complete export, while the suggested Buffer slice corrupts it. I also prepended that Unicode comment to the actual `state.js`, verified the extracted export is identical to the original, and successfully imported `evalReactComponent`. Keeping the current string slicing and resolving this as a false positive.

File: docs/app/reflex_docs/templates/docpage/docpage.py [hunk 9]
```diff
@@ -182,6 +182,54 @@ def docpage_footer(path: rx.Var[str], edit_href: rx.Var[str]) -> rx.Component:
...
+    """Build structured breadcrumbs using the visible navigation's route resolver.
+
+    Args:
+        path: The app-relative documentation path.
+        title: The current page's name.
+
+    Returns:
+        A schema.org BreadcrumbList with canonical public URLs.
+    """
+    base = "https://reflex.dev/docs"
```
Comment (masenf) [inline docs/app/reflex_docs/templates/docpage/docpage.py]: we should use `get_config().deploy_url` here so the breadcrumbs work transparently on localhost/dev/staging

File: docs/app/reflex_docs/reflex_docs.py [hunk 10]
```diff
@@ -188,21 +201,48 @@ def _canonical_url(path: str) -> str:
...
+
+app.api_transformer = partial(
+    DocsRedirectMiddleware,
+    redirects=[
+        (source, target)
+        for source, target in redirects
+        if _check_whitelisted_path(target)
+    ],
+    frontend_path=_FRONTEND_PATH,
+)
```
Comment (masenf) [inline docs/app/reflex_docs/reflex_docs.py:251]: probably fine... but causes the redirects to not work as expected in dev mode, because they depend on the request hitting the backend in order for the redirect to be serviced. So any deployment where the frontend and backend are on different origins won't really work.

File: docs/app/tests/frontend_quality.test.mjs [hunk 11]
```diff
@@ -0,0 +1,198 @@
+/** Run after `reflex export`: node --test tests/frontend_quality.test.mjs. */
```
Comment (masenf) [inline docs/app/tests/frontend_quality.test.mjs:1]: this should probably be in the CI job that runs the docs, otherwise it's just not going to get executed often if ever

File: docs/app/reflex_docs/templates/docpage/docpage.py (hunk 9 above)
Comment (Alek99) [reply docs/app/reflex_docs/templates/docpage/docpage.py]: Fixed in f2a0f434. Breadcrumbs now use the existing `public_url()` helper, which reads `get_config().deploy_url` and `frontend_path`. Canonical links use the same helper to keep their origin and mount consistent. The regression failed for localhost, staging/custom mounts, and hosting at the origin root before the fix; all cases now pass. I also imported the actual app with `REFLEX_DEPLOY_URL=https://staging.example.com` and `REFLEX_FRONTEND_PATH=/preview/docs` and verified that both canonical and breadcrumb URLs use that deployment.

File: docs/app/reflex_docs/reflex_docs.py (hunk 10 above)
Comment (Alek99) [reply docs/app/reflex_docs/reflex_docs.py:251]: Agreed: HTTP 301 is conditional on the page request reaching this backend. The redirect pages still retain `on_load=rx.redirect(target)`, and their prerendered HTML includes an immediate meta refresh and destination link. I verified `/docs/ai/features/ide/` reaches `/docs/ai/features/editor-modes/` both on the dev frontend at port 3000 (backend at 8001) and on an independent static HTTP server serving the production output. The static response is 200, not 301. f2a0f434 clarifies this boundary at the middleware registration and in the README: separate frontend/CDN deployments need matching edge redirect rules for HTTP 301 semantics; the page fallback only provides navigation. The PR description now makes that condition explicit too.

File: docs/app/tests/frontend_quality.test.mjs (hunk 11 above)
Comment (Alek99) [reply docs/app/tests/frontend_quality.test.mjs:1]: Added in f2a0f434. Both `reflex-docs` integration CI jobs (Python 3.11 and 3.12) now run `uv run --active --no-sync node --test tests/frontend_quality.test.mjs` from `docs/app` after the production build. That gives the tests the generated frontend dependencies they need and fails CI on a regression. The same command passes all four tests locally; the complete docs suite passes 445 tests (1 skipped).

File: docs/app/rxconfig.py [hunk 12]
```diff
@@ -6,6 +6,8 @@
 config = rx.Config(
     app_name="reflex_docs",
     frontend_path="/docs",
+    deploy_url="https://reflex.dev",
```
Comment (masenf) [inline docs/app/rxconfig.py]: if this is set in the rxconfig.py, then local runs will have the wrong public_url set. this value is intended to be set in the environment via REFLEX_DEPLOY_URL by the deploy job, depending on where that particular deploy will be hosting. By default it will remain localhost (as it should)

**Review by masenf — COMMENTED** [review]
not new, but i noticed that the casing for APIs/Apis and URLs/Urls looks a little off on the AI sidebar

Image

File: docs/app/rxconfig.py (hunk 12 above)
Comment (Alek99) [reply docs/app/rxconfig.py]: Fixed in a7e412af. Removed `deploy_url` from `rxconfig.py`, so local runs retain the framework localhost default and deployment jobs supply their origin through `REFLEX_DEPLOY_URL`. The README now states that explicitly. Regression coverage verifies both the unset environment (localhost) and a staging override; the complete docs suite passes 449 tests, with 1 skipped.

**Comment by Alek99** [comment]
Addressed [Masen’s AI sidebar casing feedback](https://github.com/reflex-dev/reflex/pull/7078#pullrequestreview-5172758873) in a7e412af: the APIs and URLs page titles now preserve their acronyms, and the sidebar normalizer preserves those plural forms too. The group labels and child entries now match (`APIs` / `APIs`, `URLs` / `URLs`). Both sidebar regressions failed before the fix and now pass. The complete docs suite passes 449 tests (1 skipped), and all applicable pre-commit checks pass.

**Review by masenf — APPROVED** [review]
